### PR TITLE
[Sketcher] Enable customization of edges pattern and width by type.

### PR DESCRIPTION
**Pull request by PaddleStroke**
Enable customization of edges pattern and width by type. WIP

ie adds preferences for curve pattern and width for : normal / construction / internal / external geometries.

The main goal is to enable dashed lines for construction curves, which is a long standing request:
https://forum.freecad.org/viewtopic.php?t=76435
https://forum.freecad.org/viewtopic.php?t=80099
https://forum.freecad.org/viewtopic.php?t=30217
https://github.com/FreeCAD/FreeCAD/issues/10100
etc.

The way to do it is to introduce 4 coin layers for each user layers (Abdullah started to implement multi user-layers management even if it is almost unused).

Preferences : 
image
image

Currently default is : 
- Solid lines for normal geos.
- dashed lines for construction, internal and external. 
- 3 line width for normal
- 2 line width for construction, internal and external. 

@obelisk79 it would be good that the design group discuss the default for each line type. Also note that the pattern can be changed if necessary.

Useful links for reference: https://forum.freecad.org/viewtopic.php?style=4&t=61196

### CODE REVIEW

**Comment by PaddleStroke** [comment]
Huge thanks to @kadet1090 who found the crash that held me the whole day poking in the dark !
It is now almost ready, we just need to discuss the default first.

**Comment by obelisk79** [comment]
My proposal for defaults are as follows, along with some feedback regarding the spinboxes you've added:
https://media.discordapp.net/attachments/1162044910535450665/1197513543797448744/image.png

Which results in an appearance as seen here:
https://media.discordapp.net/attachments/1162044910535450665/1197514028126326915/image.png

**Comment by pierreporte** [comment]
The preference page should be renamed since it’s not only about colors anymore. I suggest “appearance”.

File: src/Mod/Sketcher/Gui/EditModeCoinManagerParameters.h [hunk 1]
```diff
@@ -404,17 +469,62 @@ struct CoinMapping
...
+        for (size_t l = 0; l < CurvIdToGeoId.size(); ++l) {
+            CurvIdToGeoId[l].clear();
+        }
         CurvIdToGeoId.clear();
         PointIdToGeoId.clear();
         GeoElementId2SetId.clear();
         PointIdToVertexId.clear();
     };
 
+    /* //Debug function.
```
Comment (kadet1090) [inline src/Mod/Sketcher/Gui/EditModeCoinManagerParameters.h:481]: should be deleted

File: src/Mod/Sketcher/Gui/EditModeCoinManagerParameters.h [hunk 2]
```diff
@@ -461,7 +571,8 @@ struct CoinMapping
 
     //* These map a MF index (second index) within a coin layer (first index) for points or curves
     // to a GeoId */
-    std::vector<std::vector<int>> CurvIdToGeoId;   // conversion of SoLineSet index to GeoId
+    std::vector<std::vector<std::vector<int>>>
```
Comment (kadet1090) [inline src/Mod/Sketcher/Gui/EditModeCoinManagerParameters.h:574]: It is not a change for this PR, but this is getting way too complex. In that point we probably should introduce some proper structures for the layers with proper names, methods etc which would be easier to follow later. 

API could look something like that (just a concept):
```
layer(0).type(SubLayer::External).curveToGeo(curveIndex)
```

Now we have to deal with 3-level deep structures which are not easy to read and remember, stuff like `swcs[l * 4 + t]`.

File: src/Mod/Sketcher/Gui/EditModeGeometryCoinManager.cpp [hunk 3]
```diff
@@ -73,19 +73,20 @@ void EditModeGeometryCoinManager::processGeometry(const GeoListFacade& geolistfa
...
-
     auto layersconfigurations = viewProvider.VisualLayerList.getValues();
 
     for (auto l = 0; l < geometryLayerParameters.getCoinLayerCount(); l++) {
-        setEnableLayer(l, layersconfigurations[l].isVisible());
+        auto enabled = layersconfigurations[l].isVisible();
+
+        swsp[l] = enabled;
+        for (int t = 0; t < geometryLayerParameters.getSubLayerCount(); t++) {
+            swsc[l * 4 + t] = enabled;
```
Comment (kadet1090) [inline src/Mod/Sketcher/Gui/EditModeGeometryCoinManager.cpp:88]: `4` should be replaced with `getSubLayerCount()`

**Review by kadet1090 — CHANGES_REQUESTED** [review]

**Comment by PaddleStroke** [comment]
Changed name of pref page to appearance

**Review by kadet1090 — APPROVED** [review]

**Review by chennes — CHANGES_REQUESTED** [review]
Overall this looks good, thanks! The only issue I see right now is the excessive width of the Appearance tab -- can you tighten up the four lines that have the line style setting on them? On a Mac with the "Classic" theme selected the dialog is quite a bit too wide, requiring a horizontal scroll bar at the default width.

**Comment by PaddleStroke** [comment]
@chennes width of the dialog is fixed, everything fit in : 
image

**Comment by chennes** [comment]
You've still got the horizontal scrollbar there -- something is still amiss with the column widths, there is plenty of space for the widgets so I don't know why Qt wants to add it.

**Comment by PaddleStroke** [comment]
image

Width fixed so that we don't have the horizontal bar anymore.

**Pull request merged**